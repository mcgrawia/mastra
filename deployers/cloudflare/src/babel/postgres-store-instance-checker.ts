import type { NodePath, PluginObj } from '@babel/core';
import * as babel from '@babel/core';
import type { NewExpression } from '@babel/types';

/**
 * Babel plugin that enforces singleton PostgresStore instances in Cloudflare Workers.
 *
 * This plugin:
 * 1. Scans for all `new PostgresStore()` instantiations
 * 2. Records their file locations
 * 3. Throws an error if multiple instances are found
 *
 * Cloudflare Workers should only create one PostgresStore instance to avoid connection
 * pool exhaustion and ensure proper resource management.
 *
 * @returns {PluginObj} A Babel plugin object that validates PostgresStore usage
 *
 * @example
 * // Throws error if multiple instances found:
 * const store1 = new PostgresStore();
 * const store2 = new PostgresStore(); // Error thrown here
 */
export function postgresStoreInstanceChecker(): PluginObj {
  const t = babel.types;
  const instances: { path: NodePath<NewExpression>; location: string }[] = [];

  return {
    name: 'postgresstore-instance-checker',
    visitor: {
      NewExpression(path, state) {
        if (t.isIdentifier(path.node.callee) && path.node.callee.name === 'PostgresStore') {
          const filename = state.file?.opts?.filename || 'unknown file';
          const location = path.node.loc
            ? `${filename}: line ${path.node.loc.start.line}, column ${path.node.loc.start.column}`
            : 'unknown location';

          instances.push({
            path,
            location,
          });
        }
      },
    },
    post() {
      if (instances.length > 1) {
        const errorMessage = [
          `Found ${instances.length} PostgresStore instantiations:`,
          ...instances.map((instance, i) => `  ${i + 1}. At ${instance.location}`),
          'Only one PostgresStore instance should be created per Cloudflare Worker.',
        ].join('\n');

        const lastInstance = instances[instances.length - 1];
        throw lastInstance?.path.buildCodeFrameError(errorMessage);
      }
    },
  } as PluginObj;
}
