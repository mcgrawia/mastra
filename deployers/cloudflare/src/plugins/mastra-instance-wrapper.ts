import { transformSync } from '@babel/core';
import type { Plugin } from 'rollup';
import { mastraInstanceWrapper as mastraInstanceWrapperBabel } from '../babel/mastra-instance-wrapper';

export function mastraInstanceWrapper(mastraEntryFile: string): Plugin {
  return {
    name: 'mastra-wrapper',
    transform(code, id) {
      if (id !== mastraEntryFile) {
        return null;
      }

      const result = transformSync(code, {
        filename: id,
        babelrc: false,
        configFile: false,
        plugins: [mastraInstanceWrapperBabel],
      });

      if (!result?.code) {
        throw new Error('mastra-wrapper plugin did not return code, there is likely a bug in the plugin.');
      }

      return {
        code: result.code,
        map: result?.map,
      };
    },
  };
}
