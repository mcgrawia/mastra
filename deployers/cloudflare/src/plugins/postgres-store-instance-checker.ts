import { transformSync } from '@babel/core';
import type { Plugin } from 'rollup';
import { postgresStoreInstanceChecker as postgresStoreInstanceCheckerBabel } from '../babel/postgres-store-instance-checker';

export function postgresStoreInstanceChecker(): Plugin {
  return {
    name: 'postgres-store-instance-checker',
    transform(code, id) {
      const result = transformSync(code, {
        filename: id,
        babelrc: false,
        configFile: false,
        plugins: [postgresStoreInstanceCheckerBabel],
      });

      if (!result?.code) {
        throw new Error(
          'postgres-store-instance-checker plugin did not return code, there is likely a bug in the plugin.',
        );
      }

      return {
        code: result.code,
        map: result?.map,
      };
    },
  };
}
