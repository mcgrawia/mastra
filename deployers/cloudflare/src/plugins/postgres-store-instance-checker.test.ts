import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { rollup } from 'rollup';
import esbuild from 'rollup-plugin-esbuild';
import { describe, it, expect } from 'vitest';
import { postgresStoreInstanceChecker } from '../plugins/postgres-store-instance-checker';

describe('Check PostgresStore instance', () => {
  const _dirname = dirname(fileURLToPath(import.meta.url));

  it('should bundle without error', async () => {
    const file = join(_dirname, './__fixtures__/basic.js');

    const bundle = await rollup({
      logLevel: 'silent',
      input: file,
      cache: false,
      treeshake: 'smallest',
      plugins: [
        {
          name: 'externalize-all',
          resolveId(id) {
            return {
              id,
              external: id !== file,
            };
          },
        },
        postgresStoreInstanceChecker(),
        esbuild({
          target: `esnext`,
          platform: 'browser',
          minify: false,
        }),
      ],
    });

    const result = await bundle.generate({
      format: 'esm',
    });

    expect(result?.output[0].code).toMatchSnapshot();
  });

  it('should not bundle and throw an error', async () => {
    const file = join(_dirname, './__fixtures__/multiple-postgres-stores.js');

    await expect(
      async () =>
        await rollup({
          logLevel: 'silent',
          input: file,
          cache: false,
          treeshake: 'smallest',
          plugins: [
            {
              name: 'externalize-all',
              resolveId(id) {
                return {
                  id,
                  external: id !== file,
                };
              },
            },
            postgresStoreInstanceChecker(),
            esbuild({
              target: `esnext`,
              platform: 'browser',
              minify: false,
            }),
          ],
        }),
    ).rejects.toThrow(/Only one PostgresStore instance should be created per Cloudflare Worker./);
  });
});
