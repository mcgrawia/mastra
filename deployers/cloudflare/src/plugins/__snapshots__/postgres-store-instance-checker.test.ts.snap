// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Check PostgresStore instance > should bundle without error 1`] = `
"import { Mastra } from '@mastra/core/mastra';
import { createLogger } from '@mastra/core/logger';
import { weatherAgent } from '@/agents';
import { TestDeployer } from '@mastra/deployer/test';
import { createApiRoute } from '@mastra/core/server';
import { PostgresStore } from '@mastra/pg';

const mastra = new Mastra({
  agents: {
    weatherAgent
  },
  storage: new PostgresStore({
    connectionString: process.env.DATABASE_URL
  }),
  logger: createLogger({
    name: "Mastra",
    level: "info"
  }),
  telemetry: {
    enabled: true,
    serviceName: "my-app",
    export: {
      type: "otlp",
      endpoint: "http://localhost:4318"
      // SigNoz local endpoint
    }
  },
  server: {
    port: 3e3,
    timeout: 5e3,
    apiRoutes: [createApiRoute({
      path: "/hello",
      method: "get",
      handler: async (req, res) => {
        res.send("Hello World");
      }
    })]
  },
  deployer: new TestDeployer()
});

export { mastra };
"
`;
